### PR TITLE
Adds plugin mixin to "locate" items

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR adds yet another plugin mixin class, this time to allow "location" of stock items / locations.

Ref: https://github.com/inventree/InvenTree/discussions/2678

Closes https://github.com/inventree/InvenTree/issues/2845

App Integration: https://github.com/inventree/inventree-app/pull/124

This would allow, as a first class feature, development of external plugins to identify / locate / retrieve stock items or locations.

For example, flashing an LED at the location, or retrieving the stock item from a robotic storage system.

This PR also allows native integration of this feature with the app!

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/2957"></a>

### Web Integration

image

### App Integration

image

### TODO

- [x] Add JSON field to StockLocation model for use by plugins
- [ ] Documentation

### CODE REVIEW

**Comment by SchrodingersGat** [comment]
What do you mean by a "position reference" here?

**Review by matmair — APPROVED** [review]

**Comment by SchrodingersGat** [comment]
@matmair yeah, that makes total sense. 

So, add a "reference" field to the `StockLocation` model? Could be a JSON field, perhaps?

**Comment by matmair** [comment]
@SchrodingersGat yes that would be good. Maybe referenced to the plugin slug.

**Review by matmair — APPROVED** [review]

**Comment by matmair** [comment]
@SchrodingersGat this will need some updates with the plugin refactor. Ping me if you want pointers / a PR

**Comment by SchrodingersGat** [comment]
@matmair now that https://github.com/inventree/InvenTree/pull/2971 has merged I've got some conflicts to fix, working on that now.

Regarding the custom field for plugin access - I was thinking a "metadata" JSON field on the `StockLocation` model (which would not be used be internal application logic) could be provided solely for plugins to store custom data against a location. However, this would mean that each plugin would have to take care not to "clash" with whatever data it stores in that field.

Alternatively, we could have a new table which associates metadata between plugins and other tables, e.g. stock location. Then we'd want to expand this to other internal tables, such as the "part" table or "Company" table.

I can see this getting out of hand pretty quickly, and the simpler approach would be to have a single JSON field per location / per part / per customer / etc.

What are your thoughts on this one?

**Comment by matmair** [comment]
@SchrodingersGat I would recommend moving the new API endpoints and their tests to /plugin/api.py

The JSON field seems to be much simpler. If we provide a helper function to access the data that uses the plugin slug it might help with collisions.

File: InvenTree/plugin/samples/integration/locate_sample.py [hunk 1]
```diff
@@ -6,18 +6,22 @@
 
 import logging
```
Comment (matmair) [inline InvenTree/plugin/samples/integration/locate_sample.py:7]: I would recommend moving this to InvenTree/plugin/samples/locate/locate_sample.py

**Pull request merged**